File: packages/provide-uterm-server/src/provide/uterm/cli/audit.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""``uterm audit`` — verify a tamper-evident WORM audit log.

Example::

    uterm audit verify /var/log/uterm/audit.log
    uterm audit verify /var/log/uterm/audit.log --expected-seq 42 --expected-hash <hex>
"""

from __future__ import annotations

import sys
from typing import TYPE_CHECKING, Any, cast

from provide.uterm.server.audit_chain import verify_audit_log

if TYPE_CHECKING:
    import argparse


def _cmd_audit_verify(args: argparse.Namespace) -> None:
    """Execute ``uterm audit verify`` — verify the chain and exit 0/1."""
    expected_seq = getattr(args, "expected_seq", None)
    expected_hash = getattr(args, "expected_hash", None)

    # --expected-seq and --expected-hash must be supplied together (a head is a
    # (seq, hash) pair); one without the other is a usage error.
    if (expected_seq is None) != (expected_hash is None):
        args._parser.error("--expected-seq and --expected-hash must be given together")

    # The XOR guard above ensures both are set together; getattr loses that to Any.
    expected_head = cast("tuple[int, str] | None", (expected_seq, expected_hash) if expected_seq is not None else None)
    result = verify_audit_log(args.path, expected_head=expected_head)

    if result.ok:
        print(f"OK: {result.count} records, head seq={result.head_seq} hash={result.head_hash}")
        sys.exit(0)

    print(f"TAMPERED: {result.reason} at seq={result.first_bad_seq}")
    sys.exit(1)


def add_audit_subcommand(sub: Any) -> None:
    """Register the ``audit`` subcommand (with a nested ``verify`` action)."""
    audit_p = sub.add_parser(
        "audit",
        help="verify a tamper-evident WORM audit log",
        description="Verify the integrity of a hash-chained append-only audit log.",
    )
    audit_sub = audit_p.add_subparsers(dest="audit_command", metavar="ACTION")
    audit_sub.required = True

    verify_p = audit_sub.add_parser(
        "verify",
        help="verify the hash chain of an audit log file",
        description="Walk the audit log and confirm no record was inserted, deleted, reordered, or altered.",
    )
    verify_p.add_argument(
        "path",
        metavar="PATH",
        help="path to the JSONL audit log to verify",
    )
    verify_p.add_argument(
        "--expected-seq",
        type=int,
        metavar="N",
        default=None,
        help="expected head sequence number (requires --expected-hash)",
    )
    verify_p.add_argument(
        "--expected-hash",
        metavar="HEX",
        default=None,
        help="expected head record hash (requires --expected-seq)",
    )
    # Stash the parser so the handler can emit a proper argparse usage error
    # (exit code 2) when exactly one of the --expected-* pair is supplied.
    verify_p.set_defaults(func=_cmd_audit_verify, _parser=verify_p)

File: packages/provide-uterm-go/cli/audit.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package cli

import (
	"fmt"
	"io"

	"github.com/spf13/cobra"
)

// newAuditCmd mirrors the Python `audit` subcommand and its nested `verify`
// action — a byte-exact port of the tamper-evident WORM hash-chain verifier.
func newAuditCmd() *cobra.Command {
	audit := &cobra.Command{
		Use:   "audit",
		Short: "verify a tamper-evident WORM audit log",
		Long:  "Verify the integrity of a hash-chained append-only audit log.",
	}
	verify := &cobra.Command{
		Use:          "verify PATH",
		Short:        "verify the hash chain of an audit log file",
		Long:         "Walk the audit log and confirm no record was inserted, deleted, reordered, or altered.",
		Args:         cobra.ExactArgs(1),
		SilenceUsage: true,
		RunE: func(cmd *cobra.Command, args []string) error {
			f := cmd.Flags()
			var head *expectedHead
			seqSet := f.Changed("expected-seq")
			hashSet := f.Changed("expected-hash")
			// A head is a (seq, hash) pair; one without the other is a usage error.
			if seqSet != hashSet {
				return fmt.Errorf("--expected-seq and --expected-hash must be given together")
			}
			if seqSet {
				seq, _ := f.GetInt("expected-seq")
				hash, _ := f.GetString("expected-hash")
				head = &expectedHead{seq: int64(seq), hash: hash}
			}
			return runAuditVerify(args[0], head, cmd.OutOrStdout())
		},
	}
	vf := verify.Flags()
	vf.Int("expected-seq", 0, "expected head sequence number (requires --expected-hash)")
	vf.String("expected-hash", "", "expected head record hash (requires --expected-seq)")
	audit.AddCommand(verify)
	return audit
}

// runAuditVerify verifies the chain and reports OK / TAMPERED. It returns
// errTampered (a non-nil error) when the chain is broken so Execute maps it to a
// non-zero exit code, matching the Python `sys.exit(1)` contract.
func runAuditVerify(path string, head *expectedHead, out io.Writer) error {
	result := verifyAuditLog(path, head)
	if result.OK {
		_, _ = fmt.Fprintf(out, "OK: %d records, head seq=%s hash=%s\n",
			result.Count, fmtSeq(result.HeadSeq), fmtHash(result.HeadHash))
		return nil
	}
	_, _ = fmt.Fprintf(out, "TAMPERED: %s at seq=%s\n", result.Reason, fmtSeq(result.FirstBadSeq))
	return errTampered
}

// errTampered signals a verification failure without printing a second "error:"
// line — the TAMPERED report is already written to stdout.
var errTampered = &silentError{}

type silentError struct{}

func (*silentError) Error() string { return "" }

func fmtSeq(v *int64) string {
	if v == nil {
		return "None"
	}
	return fmt.Sprintf("%d", *v)
}

func fmtHash(v *string) string {
	if v == nil {
		return "None"
	}
	return *v
}
